lalrpop-snap/Cargo.toml

```diff
@@ -24,7 +24,7 @@ itertools = "0.7"
 regex = "0.2.1"
 regex-syntax = "0.4.0"
 petgraph = "0.4.4"
-string_cache = "0.7.1"
+string_cache = "0.7.1,<=0.7.1"
```

Comment:
Couldn't this be `string_cache = "=0.7.1"`?

Reply:
Locking dep to a particular version can be problematic. For example, this would make LALRPOP unusable with anything which specifies `string_cache = "0.7.2"`. I think a better solution would be to do `cargo update string_cache "0.7.1"` in CI for Rust 1.20.0, if we don't want to just bump the minimal Rust version.

Reply:
I think =0.7.1 should be fine. I agree with matklad that dependencies shouldn't be locked to a specific version, but in this case there is no direct solution unless we specify a minimum compiler version.

Reply:
>Since it is an internal dependency it is ok to have it locked down

Unfortunately Cargo is not able to distinguish between internal and public dependencies yet, so it just flatly prohibits resolution graphs with two semver-compatible versions. That is, if crate `foo` has

```
[dependencies]
string_cache = "0.7.2"
lalrpop = "0.13.2" # pretend this is the new published lalrpop version with `=` or `<=` constraint
``` 

then Cargo won't be able to resolve its dependencies at all. 


In general, I would argue that semver open `0.7.1` is the correct constraint here, regardless of the above issue. 

It is the leaf crate with Cargo.lock that ultimately decides about dependencies. If the leaf crate must be built with Rust 1.20.0, it should have `string_cache = 0.7.1` in lockfile. If it is build with say, `1.25.0`, it should have `string_cache = 0.7.2` in the lock. The job of LALRPOP here is to be compatible with both, and not to impose additional versioning constraints. 

Does this make sense? :)

Reply:
In general, we plan to lint against using non semver-open dependencies for published crates, because there's rarely a need for that: https://github.com/rust-lang/cargo/issues/5340. 

In particular, I believe that comparability with older `rustc` is also not a reason to constraint upper versions of dependencies in Cargo.toml. Instead the right predicate here is perhaps that *minimal* versions specified in Cargo.toml are compatible with desired `rustc` version, and picking particular versions should be left to the leaf crate with a lockfile, because only the leaf knows which version of rustc is actually used.

Reply:
So cargo can only use multiple versions of the same dependency if they differ in major version? (or minor for 0.x crates)?

> In general, I would argue that semver open 0.7.1 is the correct constraint here, regardless of the above issue.

That is the better compromise, though it can lead to confusing problems for end users as well if they use <1.23. Not as bad as cargo refusing to compile though.

Reply:
>So cargo can only use multiple versions of the same dependency if they differ in major version? (or minor for 0.x crates)?

That is correct! 

>though it can lead to confusing problems for end users as well if they use <1.23.

Yeah. Ideally, Cargo should be able to take current version of Rust into account during resolve, but that is non-trivial. 

Meanwhile I think we have a nightly feature (which hopefully will be stabilized soon) which should make testing rustc compat easier: https://github.com/lalrpop/lalrpop/pull/366#discussion_r183211817.

Reply:
How about getting servo to yank 0.7.2 and issue a new 0.8.0 release that will make matters easier? as this can be considered a breaking change even though BCs are allowed in 0.x land

Reply:
There’s no consensus as to whether bumping required rustc version is a breaking change or not (discussion: https://github.com/rust-lang-nursery/api-guidelines/issues/123). I personally am convinced that it is not a breaking change.

Reply:
Agree.  it is one of those "debatable" topics. but won't you consider it a breaking change considering the new release won't work for anyone running  < 1.23?

Reply:
Huh, turns out I feel strongly about it, so I've dropped a comment on that thread: https://github.com/rust-lang-nursery/api-guidelines/issues/123#issuecomment-383329604 :0)

Reply:
Uhmm makes sense. So i think the best thing to do here is to bump mvsr for lalrpop?

Reply:
@Dylan-DPC For the moment I think it is enough to just include a lockfile.